extension/src/state_aggregate.rs

```diff
@@ -248,6 +249,16 @@ struct Record {
     time: i64,
 }
 
+fn microseconds_to_interval(micros: i64) -> *const pg_sys::Interval {
```

Comment:
It took me a while to come up with this, and it finally came as basically a wild guess. I couldn't find any example of us doing this already. It does work locally, though I'm not sure whether that's by accident, nor am I certain it's not a leak...

Reply:
[Here's](https://doxygen.postgresql.org/backend_2utils_2adt_2timestamp_8c_source.html#l01497) the canonical postgres code to create an interval from a bunch of integers. I think the code you have looks correct, though it could perhaps be simplified to
```rust
let i = pg_sys::palloc0(std::mem::size_of::<pg_sys::Interval>()) as *mut pg_sys::Interval;
(*i).time = micros;
i
```
or even
```rust
let i = pg_sys::palloc0(std::mem::size_of::<pg_sys::Interval>()) as *mut pg_sys::Interval;
*i = pg_sys::Interval {
    month: 0,
    day: 0,
    time: 0,
};
i
```

Reply:
Yeah I didn't think of assigning the whole struct, which is the obvious move and let's me have rustc make sure I don't forget to set any fields.

I guess `palloc0` is analogous to `calloc` in that it allocates and zeroes, but that's not generally safe and I prefer to avoid it. That point of contention is much older than my career and I'd rather not litigate it here, just explaining why I avoided palloc0 :). The postgres code you linked to is also explicit rather than implicit :)

Anyway it's much better now: thanks!